gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskEntry.java

```diff
@@ -614,37 +652,39 @@ else if (value == Token.TOMBSTONE) {
                 bytes = ohproxy.getSerializedValue();
               } else {
                 //TODO:Asif: Speak to Darrel for cleaner way
-                if(ohproxy instanceof ByteSource) {
+                if (ohproxy instanceof ByteSource) {
                   bytes = ((ByteSource)ohproxy).getRowBytes();
-                }else {
-                  bytes = (byte[]) ohproxy.getDeserializedForReading();
+                } else {
+                  bytes = (byte[])ohproxy.getDeserializedForReading();
                 }
               }
             } else {
               bytes = proxy.getSerializedValue();
             }
+            buffer = Helper.wrapBytes(bytes);
           }
           else if (value instanceof byte[]) {
             isSerializedObject = false;
-            bytes = (byte[])value;
+            buffer = ByteBuffer.wrap((byte[])value);
           }
           else {
             Assert.assertTrue(!Token.isRemovedFromDisk(value));
-            bytes = EntryEventImpl.serialize(value);
-            if (bytes.length == 0) {
-              throw new IllegalStateException("serializing <" + value + "> produced empty byte array");
+            buffer = EntryEventImpl.serializeDirect(value, null);
```

Comment:
Yes, for off-heap storage in SNAP-1454 the value is DirectByteBuffer. That is optimized by:

a) keeping full serialized format of the off-heap value in DirectByteBuffer with data i.e. with the serialization header of typeId, number of bytes

b) BlobHelper method which will check if value is a "SerializedBufferData" then will directly send its buffer as the result

Those changes are already on branch SNAP-1454 (https://github.com/SnappyDataInc/snappy-store/blob/SNAP-1454/gemfire-core/src/main/java/com/gemstone/gemfire/internal/util/BlobHelper.java#L110) and PR for that will be up by tomorrow.